### PR TITLE
[WIP] use new dune ctypes stanza

### PR DESCRIPTION
**Pull request by mbacarella**
Ports ocaml-yaml to use the anticipated dune ctypes feature
https://github.com/ocaml/dune/pull/3905

Using this feature gets rid of the stub generator utilities and intermediate libraries.  This PR also drops some deprecated/removed dune features.

### CODE REVIEW

File: ffi/dune [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ (libraries sexplib)
+ (foreign_archives ../vendor/yaml_c)
+ (ctypes
+  (external_library_name yaml)
+  (build_flags_resolver
+   (vendored
+     ; hack: repeated include dirs from root and relative since this gets
+     ;  called from multiple places
+    (c_flags (-I../vendor/headers -Ivendor/headers))))
+  (headers (preamble "#if defined(__MINGW32__) || defined(__MINGW64__)
```
Comment (avsm) [inline ffi/dune:13]: This preamble might be better off in a separate file, and referenced from here. I can see it getting awkward fast to embed C headers in dune files.

**Comment by avsm** [comment]
This is a very pleasing amount of red indeed! The new stanzas are looking awesome...

File: ffi/dune (hunk 1 above)
Comment (mbacarella) [reply ffi/dune]: Oh, the dune stanza already supports that.  I just pushed a version that does the include approach instead.

(I think it's still nice to provide an option to include a preamble in the dune feature as well)

**Comment by mbacarella** [comment]
marked as WIP.  this depends on dune 3.0 to be released (and some additional meta file fixups) before it can be merged.

**Comment by avsm** [comment]
I haven't reviewed in detail, but the overall approach in this PR looks very good to me.

**Comment by mbacarella** [comment]
> How ready is this PR ? It will be useful for dune monorepos as there are issues with the ctypes port currently. For info Dune 3 is getting released very soon ! [ocaml/opam-repository#20721](https://github.com/ocaml/opam-repository/pull/20721)

I'm not aware of any deficiencies, it's WIP because I don't want it accidentally merged with the pin-depends to a specific dune version in there. I'll freshen this PR up after dune 3.0 is released.

File: ffi/dune [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ (public_name yaml.ffi)
+ (libraries sexplib)
+ (foreign_archives ../vendor/yaml_c)
+ (ctypes
+  (external_library_name yaml)
+  (build_flags_resolver
+   (vendored
+     ; hack: repeated include dirs from root and relative since this gets
+     ;  called from multiple places
+    (c_flags (-I../vendor/headers -Ivendor/headers))))
```
Comment (TheLortex) [inline ffi/dune:12]: I think the only issue is here: this doesn't work in a dune workspace created with opam-monorepo. 
This is already known but I'm linking the relevant issue : https://github.com/ocaml/dune/issues/5325

**Comment by sim642** [comment]
> I'll freshen this PR up after dune 3.0 is released.

That has now been released, so this PR could be finalized.

**Comment by mbacarella** [comment]
Closing in favor of https://github.com/avsm/ocaml-yaml/pull/67

**Pull request closed without merging**